### PR TITLE
adding batched autoMF

### PR DESCRIPTION
**Pull request by gganssle**
What do you think about this? Fraught with local minima? Perhaps. But will it work? I claim yes.

### CODE REVIEW

**Comment by EthanRosenthal** [comment]
Yeah, I think that'd work - it's basically classical gradient descent where you take the gradient with respect to the entire dataset. I believe it should converge slower than stochastic or mini-batch gradient descent, but you find a better minimum ([see pg. 276](http://www.deeplearningbook.org/contents/optimization.html)).

In the case of the code I wrote for my blog post, it's super slow because python's inefficient in looping, and we're not really able to exploit parallelism well on backpropagating through single gradients.

The best way to implement this all is to define a PyTorch DataLoader and then play with the batch size (trading off speed vs. error). The other nice thing about the DataLoaders is that you can load data in parallel using multiple workers, if loading data is a bottleneck. 

Here's an example adapted from my dated [torchmf](https://github.com/EthanRosenthal/torchmf) code:

```python
import numpy as np
import scipy.sparse as sp

import torch
from torch.autograd import Variable
from torch.utils import data as torchdata


class Interactions(torchdata.Dataset):

    def __init__(self, mat):
        mat = mat.astype(np.float32).tocoo()
        self.row = torch.from_numpy(mat.row)
        self.col = torch.from_numpy(mat.col)
        self.data = torch.from_numpy(mat.data)
        self.n_users = mat.shape[0]
        self.n_items = mat.shape[1]

    def __getitem__(self, index):
        row = self.row[index]
        col = self.col[index]
        val = self.data[index]
        return (row, col), val

    def __len__(self):
        return len(self.data)


class MatrixFactorization(torch.nn.Module):
    
    def __init__(self, n_users, n_items, n_factors=20):
        super().__init__()
        self.user_factors = torch.nn.Embedding(n_users, 
                                               n_factors)
        self.item_factors = torch.nn.Embedding(n_items, 
                                               n_factors)
        
    def forward(self, user, item):
        return (self.user_factors(user) * self.item_factors(item)).sum(1)

n_users = 1000
n_items = 1000
ratings = sp.rand(n_users, n_items, 
                  density=0.01, format='csr')
ratings.data = (np.random.randint(1, 5, 
                                  size=ratings.nnz)
                          .astype(np.float64))

model = MatrixFactorization(n_users, n_items)
loss_func = torch.nn.MSELoss()
optimizer = torch.optim.SGD(model.parameters(), 
                            lr=1e-6)

batch_size = 256
num_workers = 0

dataloader = torchdata.DataLoader(Interactions(ratings),
                                  batch_size=batch_size,
                                  shuffle=True,
                                  num_workers=num_workers,
                                  drop_last=False)

for (user, item), rating in dataloader:

    # Turn data into variables
    user, item, rating = Variable(user), Variable(item), Variable(rating.float())

    # Predict and calculate loss
    prediction = model(user, item)
    loss = loss_func(prediction, rating)
    
    optimizer.zero_grad()

    # Backpropagate
    loss.backward()
        
    # Update the parameters
    optimizer.step()
```

**Comment by gganssle** [comment]
Yeah, It works (and at high speed). I'm using a similar architecture to train a model on a bunch of home gift tchotchkes to build a bigger meta model for customer bucketing. Glad to see you have it implemented in the real (non-blog) code.
